cpp/src/arrow/compute/exec/expression.cc

```diff
@@ -512,7 +509,31 @@ Result<Datum> ExecuteScalarExpression(const Expression& expr, const ExecBatch& i
       return MakeNullScalar(null());
     }
 
-    const Datum& field = input[param->index];
+    Datum field = input[param->indices[0]];
```

Comment:
Minor nit:
```suggestion
    const Datum* field = &input[param->indices[0]];
```
use a non-owning reference to avoid unnecessary shared pointer thrashing

Reply:
That doesn't work so well because in the StructArray case, we're allocating a new array, and in the StructScalar case, the scalar contains Scalar values, not Datum values.